File: src/domain/garment-form.test.ts
import { describe, expect, it } from 'vitest';
import {
  EMPTY_FORM,
  brandSuggestions,
  displayedPreviewUri,
  galleryItems,
  normalizeForm,
  selectedHasOriginal,
  toggled,
  withBackgroundRemoved,
  withDetectedColor,
  withImage,
  withImagesReordered,
  withImportedPreview,
  withSubcategories,
  withoutImageAt,
  type GarmentFormState,
} from './garment-form';
import type { SeasonOption } from '../constants/style-filters';

const form = (overrides: Partial<GarmentFormState> = {}): GarmentFormState =>
  normalizeForm({ ...overrides });

// The invariant behind most of this: bgRemovedUris is positionally aligned with
// imageUris, so entry i is the cut-out of photo i.
const aligned = (state: GarmentFormState) =>
  state.bgRemovedUris.length === state.imageUris.length;

describe('normalizeForm', () => {
  it('aligns the cut-out list with the photos', () => {
    const state = normalizeForm({ imageUris: ['a', 'b', 'c'], bgRemovedUris: ['a-nobg'] });

    expect(state.bgRemovedUris).toEqual(['a-nobg', '', '']);
    expect(aligned(state)).toBe(true);
  });

  it('trims a cut-out list longer than the photos', () => {
    const state = normalizeForm({ imageUris: ['a'], bgRemovedUris: ['x', 'y', 'z'] });
    expect(state.bgRemovedUris).toEqual(['x']);
  });

  it('falls back to a default palette rather than an empty one', () => {
    expect(normalizeForm({ colorPalette: [] }).colorPalette).toEqual(['#000000']);
    expect(normalizeForm({ colorPalette: ['#FFFFFF'] }).colorPalette).toEqual(['#FFFFFF']);
  });
});

describe('withImage', () => {
  it('appends and selects the new photo', () => {
    const state = withImage(form({ imageUris: ['a'], bgRemovedUris: ['a-nobg'] }), 'b');

    expect(state.imageUris).toEqual(['a', 'b']);
    expect(state.bgRemovedUris).toEqual(['a-nobg', '']);
    expect(state.selectedImageIndex).toBe(1);
    expect(aligned(state)).toBe(true);
  });

  it('replacing the selected photo drops its cut-out', () => {
    // The old cut-out was of the old photo, so keeping it would show the wrong
    // garment.
    const before = form({ imageUris: ['a', 'b'], bgRemovedUris: ['a-nobg', 'b-nobg'], selectedImageIndex: 1 });
    const state = withImage(before, 'b2', true);

    expect(state.imageUris).toEqual(['a', 'b2']);
    expect(state.bgRemovedUris).toEqual(['a-nobg', '']);
    expect(state.selectedImageIndex).toBe(1);
  });

  it('appends when asked to replace but there is nothing selected', () => {
    const state = withImage(form(), 'a', true);
    expect(state.imageUris).toEqual(['a']);
  });
});

describe('withoutImageAt', () => {
  it('removes the photo and its cut-out together', () => {
    const state = withoutImageAt(
      form({ imageUris: ['a', 'b', 'c'], bgRemovedUris: ['a1', 'b1', 'c1'] }),
      1
    );

    expect(state.imageUris).toEqual(['a', 'c']);
    expect(state.bgRemovedUris).toEqual(['a1', 'c1']);
    expect(aligned(state)).toBe(true);
  });

  it('shifts the selection down when an earlier photo goes', () => {
    const state = withoutImageAt(
      form({ imageUris: ['a', 'b', 'c'], selectedImageIndex: 2 }),
      0
    );
    // Still pointing at 'c'.
    expect(state.imageUris[state.selectedImageIndex]).toBe('c');
  });

  it('keeps the selection in range when the last photo goes', () => {
    const state = withoutImageAt(
      form({ imageUris: ['a', 'b'], selectedImageIndex: 1 }),
      1
    );
    expect(state.selectedImageIndex).toBe(0);
    expect(state.imageUris[state.selectedImageIndex]).toBe('a');
  });

  it('resets the selection when nothing is left', () => {
    const state = withoutImageAt(form({ imageUris: ['a'], selectedImageIndex: 0 }), 0);
    expect(state.imageUris).toEqual([]);
    expect(state.selectedImageIndex).toBe(0);
  });

  it('never leaves the selection past the end', () => {
    // Whichever photo goes, the index must still address something.
    for (const removed of [0, 1, 2]) {
      for (const selected of [0, 1, 2]) {
        const state = withoutImageAt(
          form({ imageUris: ['a', 'b', 'c'], selectedImageIndex: selected }),
          removed
        );
        expect(state.selectedImageIndex).toBeLessThan(state.imageUris.length);
        expect(aligned(state)).toBe(true);
      }
    }
  });
});

describe('withImagesReordered', () => {
  it('moves the photo and its cut-out together, and follows it', () => {
    const state = withImagesReordered(
      form({ imageUris: ['a', 'b', 'c'], bgRemovedUris: ['a1', 'b1', 'c1'] }),
      0,
      2
    );

    expect(state.imageUris).toEqual(['b', 'c', 'a']);
    expect(state.bgRemovedUris).toEqual(['b1', 'c1', 'a1']);
    expect(state.selectedImageIndex).toBe(2);
  });

  it('does nothing for a no-op or an out-of-range move', () => {
    const before = form({ imageUris: ['a', 'b'] });

    expect(withImagesReordered(before, 1, 1)).toBe(before);
    expect(withImagesReordered(before, 0, 5)).toBe(before);
    expect(withImagesReordered(before, -1, 0)).toBe(before);
  });
});

describe('withBackgroundRemoved', () => {
  it('records the cut-out against the selected photo only', () => {
    const state = withBackgroundRemoved(
      form({ imageUris: ['a', 'b'], selectedImageIndex: 1 }),
      'b-nobg'
    );
    expect(state.bgRemovedUris).toEqual(['', 'b-nobg']);
  });
});

describe('withSubcategories', () => {
  const seasonsFor = (): SeasonOption[] => ['winter'];

  it('fills in seasons when the user has chosen none', () => {
    const state = withSubcategories(form(), ['Parka'], seasonsFor);
    expect(state.seasons).toEqual(['winter']);
  });

  it('never overwrites seasons the user chose', () => {
    const state = withSubcategories(form({ seasons: ['summer'] }), ['Parka'], seasonsFor);
    expect(state.seasons).toEqual(['summer']);
  });
});

describe('withDetectedColor', () => {
  it('puts the detection first and keeps the rest', () => {
    // A detection is a suggestion, not a correction: a deliberate choice stays.
    const state = withDetectedColor(form({ colorPalette: ['#FFFFFF', '#CC0000'] }), '#000080');
    expect(state.colorPalette).toEqual(['#000080', '#FFFFFF', '#CC0000']);
  });

  it('does not duplicate a colour already in the palette', () => {
    const state = withDetectedColor(form({ colorPalette: ['#FFFFFF', '#CC0000'] }), '#CC0000');
    expect(state.colorPalette).toEqual(['#CC0000', '#FFFFFF']);
  });
});

describe('withImportedPreview', () => {
  it('keeps a brand the user has already typed', () => {
    const state = withImportedPreview(
      form({ brand: 'Mine' }),
      { downloadedImageUris: ['a'], brand: 'Theirs' }
    );
    expect(state.brand).toBe('Mine');
  });

  it('takes the imported brand when the field is blank', () => {
    const state = withImportedPreview(
      form({ brand: '   ' }),
      { downloadedImageUris: ['a', 'b'], brand: 'Theirs' }
    );
    expect(state.brand).toBe('Theirs');
    expect(state.bgRemovedUris).toEqual(['', '']);
    expect(state.selectedImageIndex).toBe(0);
  });
});

describe('brandSuggestions', () => {
  const known = ['Uniqlo', 'Nike', 'New Balance', 'Adidas'];

  it('offers everything when nothing has been typed', () => {
    expect(brandSuggestions(known, '')).toEqual(known);
    expect(brandSuggestions(known, '   ')).toEqual(known);
  });

  it('matches case-insensitively anywhere in the name, not just the start', () => {
    // 'ni' is inside U-ni-qlo as well as starting Nike, and both are offered:
    // the match is a substring, not a prefix.
    expect(brandSuggestions(known, 'ni')).toEqual(['Uniqlo', 'Nike']);
    expect(brandSuggestions(known, 'BALANCE')).toEqual(['New Balance']);
    expect(brandSuggestions(known, 'qlo')).toEqual(['Uniqlo']);
  });

  it('drops an exact match, which the user has already typed', () => {
    expect(brandSuggestions(known, 'Nike')).toEqual([]);
    expect(brandSuggestions(known, '  nike  ')).toEqual([]);
  });

  it('caps the list', () => {
    const many = Array.from({ length: 20 }, (_, i) => `Brand${i}`);
    expect(brandSuggestions(many, '')).toHaveLength(8);
  });
});

describe('derived view values', () => {
  it('shows the cut-out where there is one, else the photo', () => {
    const state = form({ imageUris: ['a', 'b'], bgRemovedUris: ['a-nobg', ''] });

    expect(galleryItems(state)).toEqual([
      { uri: 'a-nobg', original: 'a' },
      { uri: 'b', original: 'b' },
    ]);
    expect(displayedPreviewUri(state)).toBe('a-nobg');
    expect(displayedPreviewUri({ ...state, selectedImageIndex: 1 })).toBe('b');
  });

  it('reports no preview for an empty form', () => {
    expect(displayedPreviewUri(EMPTY_FORM)).toBeNull();
    expect(galleryItems(EMPTY_FORM)).toEqual([]);
  });

  it('knows a cut-out-only garment has nothing to undo to', () => {
    // Both slots hold the same path, so there is no with-background original.
    const cutoutOnly = form({ imageUris: ['only.png'], bgRemovedUris: ['only.png'] });
    expect(selectedHasOriginal(cutoutOnly)).toBe(false);

    const hasBoth = form({ imageUris: ['a.jpg'], bgRemovedUris: ['a-nobg.png'] });
    expect(selectedHasOriginal(hasBoth)).toBe(true);
  });
});

describe('toggled', () => {
  it('adds what is absent and removes what is present', () => {
    expect(toggled(['a'], 'b')).toEqual(['a', 'b']);
    expect(toggled(['a', 'b'], 'a')).toEqual(['b']);
  });
});

File: presentation/src/test/kotlin/com/wardrobapp/presentation/GarmentFormTest.kt
package com.wardrobapp.presentation

import com.wardrobapp.domain.Season
import com.wardrobapp.domain.seasonsForSubcategories
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The rules of the add/edit form.
 *
 * Two lists have to stay aligned -- the photos and their cut-outs, entry for
 * entry -- and almost every wrong answer here loses a photo or keeps a file
 * forever. The other half is about not overwriting what somebody typed: a
 * detected colour, an implied season and an imported brand are all suggestions,
 * and a suggestion that overwrites a choice is a bug the user watches happen.
 *
 * These cases replace the recorded transition scripts, which came from the app
 * this was ported from.
 */
class GarmentFormTest {

    private fun form(
        images: List<String> = emptyList(),
        cutouts: List<String> = emptyList(),
        selected: Int = 0,
    ) = GarmentFormState(
        imageUris = images,
        bgRemovedUris = cutouts,
        selectedImageIndex = selected,
    ).normalized()

    @Test
    fun `normalizing aligns the cut-outs with the photos and keeps a colour`() {
        val state = GarmentFormState(
            imageUris = listOf("a.jpg", "b.jpg", "c.jpg"),
            bgRemovedUris = listOf("a-cut.png"),
            colorPalette = emptyList(),
        ).normalized()

        assertEquals(3, state.bgRemovedUris.size)
        assertEquals(listOf("a-cut.png", "", ""), state.bgRemovedUris)
        assertEquals(listOf(GarmentFormState.DEFAULT_COLOR), state.colorPalette)

        // Extra cut-outs are dropped rather than left dangling past the end.
        val trimmed = GarmentFormState(
            imageUris = listOf("a.jpg"),
            bgRemovedUris = listOf("a-cut.png", "orphan.png"),
        ).normalized()
        assertEquals(listOf("a-cut.png"), trimmed.bgRemovedUris)
    }

    @Test
    fun `adding a photo selects it, and both lists grow together`() {
        val state = form(listOf("a.jpg")).withImage("b.jpg")

        assertEquals(listOf("a.jpg", "b.jpg"), state.imageUris)
        assertEquals(listOf("", ""), state.bgRemovedUris)
        assertEquals(1, state.selectedImageIndex)
    }

    @Test
    fun `replacing a photo drops the cut-out that was made of it`() {
        // The cut-out was of the old photo; keeping it would show one garment's
        // outline over another's picture.
        val state = form(listOf("a.jpg", "b.jpg"), listOf("a-cut.png", "b-cut.png"), selected = 0)
            .withImage("new.jpg", replaceCurrent = true)

        assertEquals(listOf("new.jpg", "b.jpg"), state.imageUris)
        assertEquals(listOf("", "b-cut.png"), state.bgRemovedUris)
    }

    @Test
    fun `replacing when there is nothing there adds instead`() {
        val state = form().withImage("first.jpg", replaceCurrent = true)

        assertEquals(listOf("first.jpg"), state.imageUris)
        assertEquals(listOf(""), state.bgRemovedUris)
    }

    @Test
    fun `removing a photo keeps the selection pointing at something`() {
        val three = form(listOf("a.jpg", "b.jpg", "c.jpg"), listOf("", "b-cut.png", ""), selected = 2)

        // Removing something before the selection shifts it back with the list.
        val removedFirst = three.withoutImageAt(0)
        assertEquals(listOf("b.jpg", "c.jpg"), removedFirst.imageUris)
        assertEquals(listOf("b-cut.png", ""), removedFirst.bgRemovedUris)
        assertEquals(1, removedFirst.selectedImageIndex)

        // Removing the selection itself lands on what is now in that place, or
        // the end of the list.
        val removedLast = three.withoutImageAt(2)
        assertEquals(1, removedLast.selectedImageIndex)

        // Removing the only photo leaves a valid, empty form.
        val emptied = form(listOf("a.jpg")).withoutImageAt(0)
        assertEquals(emptyList(), emptied.imageUris)
        assertEquals(0, emptied.selectedImageIndex)
    }

    @Test
    fun `reordering carries the cut-out and follows with the selection`() {
        val state = form(listOf("a.jpg", "b.jpg", "c.jpg"), listOf("", "b-cut.png", ""))
            .withImagesReordered(fromIndex = 1, toIndex = 0)

        assertEquals(listOf("b.jpg", "a.jpg", "c.jpg"), state.imageUris)
        assertEquals(listOf("b-cut.png", "", ""), state.bgRemovedUris)
        assertEquals(0, state.selectedImageIndex)
    }

    @Test
    fun `a reorder that goes nowhere changes nothing`() {
        val state = form(listOf("a.jpg", "b.jpg"))

        assertEquals(state, state.withImagesReordered(1, 1))
        assertEquals(state, state.withImagesReordered(0, 5))
        assertEquals(state, state.withImagesReordered(-1, 0))
    }

    @Test
    fun `a cut-out is recorded against the photo it was made from`() {
        val state = form(listOf("a.jpg", "b.jpg"), selected = 1).withBackgroundRemoved("b-cut.png")

        assertEquals(listOf("", "b-cut.png"), state.bgRemovedUris)

        // And clearing it is the same call with nothing.
        assertEquals(listOf("", ""), state.withBackgroundRemoved("").bgRemovedUris)
    }

    @Test
    fun `collapsing stores the cut-out and hands back the original to delete`() {
        // Saving space is the point of removing a background, so keeping both
        // files would make every removal cost storage rather than save it.
        val stored = form(listOf("a.jpg", "b.jpg"), listOf("a-cut.png", "")).imagesToStore()

        assertEquals(listOf("a-cut.png", "b.jpg"), stored.imageUris)
        assertEquals(listOf("a-cut.png", ""), stored.bgRemovedUris)
        assertEquals(listOf("a.jpg"), stored.discardable)
    }

    @Test
    fun `collapsing an already-collapsed garment finds nothing to discard`() {
        // Stated directly because it needs the collapse fed back in, which is
        // what editing a saved garment does. Getting it wrong deletes the photo
        // the garment is showing.
        val once = GarmentFormState(imageUris = listOf("a.jpg"), bgRemovedUris = listOf("a-cut.png"))
            .normalized()
            .imagesToStore()

        val twice = GarmentFormState(
            imageUris = once.imageUris,
            bgRemovedUris = once.bgRemovedUris,
        ).normalized().imagesToStore()

        assertEquals(listOf("a-cut.png"), twice.imageUris)
        assertEquals(listOf("a-cut.png"), twice.bgRemovedUris)
        assertEquals(emptyList(), twice.discardable)
    }

    @Test
    fun `a garment type fills the seasons in, but never over a choice`() {
        val implied = form().withSubcategories(listOf("Sandals"), ::seasonsForSubcategories)
        assertEquals(listOf(Season.SUMMER), implied.seasons)

        val chosen = form().copy(seasons = listOf(Season.WINTER))
            .withSubcategories(listOf("Sandals"), ::seasonsForSubcategories)
        assertEquals(listOf(Season.WINTER), chosen.seasons, "an explicit choice was overwritten")
    }

    @Test
    fun `a garment always has a colour`() {
        val one = form().copy(colorPalette = listOf("#FF0000"))

        // Toggling the last colour off puts the default back rather than leaving
        // the palette empty.
        assertEquals(
            listOf(GarmentFormState.DEFAULT_COLOR),
            one.withColorToggled("#FF0000").colorPalette,
        )
        // Toggling is add-if-absent, remove-if-present.
        assertEquals(listOf("#FF0000", "#00FF00"), one.withColorToggled("#00FF00").colorPalette)
    }

    @Test
    fun `a detected colour goes first and keeps what was already picked`() {
        // A detection is a suggestion, not a correction: replacing the palette
        // would discard a deliberate choice.
        val state = form().copy(colorPalette = listOf("#FF0000", "#00FF00"))
            .withDetectedColor("#0000FF")

        assertEquals(listOf("#0000FF", "#FF0000", "#00FF00"), state.colorPalette)

        // Detecting a colour that is already there moves it up rather than
        // listing it twice.
        val again = state.withDetectedColor("#FF0000")
        assertEquals(listOf("#FF0000", "#0000FF", "#00FF00"), again.colorPalette)
    }

    @Test
    fun `a suggested type joins the ones already picked in the same category`() {
        val state = form().copy(category = "tops", subcategories = listOf("Shirt"))
            .withSuggestedType("tops", "Polo", ::seasonsForSubcategories)

        assertEquals(listOf("Polo", "Shirt"), state.subcategories)
        assertEquals("tops", state.category)

        // And suggesting one that is already picked moves it up rather than
        // listing it twice, exactly as a detected colour does.
        val again = state.withSuggestedType("tops", "Shirt", ::seasonsForSubcategories)
        assertEquals(listOf("Shirt", "Polo"), again.subcategories)
    }

    @Test
    fun `a suggestion in another category replaces the types rather than mixing them`() {
        // "Sneakers, Shirt" is not a garment. The types belonged to a category this
        // one is not, which is why tapping a category chip clears them too.
        val state = form().copy(category = "tops", subcategories = listOf("Shirt"))
            .withSuggestedType("shoes", "Sneakers", ::seasonsForSubcategories)

        assertEquals("shoes", state.category)
        assertEquals(listOf("Sneakers"), state.subcategories)
    }

    @Test
    fun `a suggestion with no type still moves the form to the category`() {
        // What the labeller reported was "Footwear", and narrowing the chips from
        // seventy types to seven is the whole value of that.
        val state = form().copy(category = "tops", subcategories = listOf("Shirt"))
            .withSuggestedType("shoes", null, ::seasonsForSubcategories)

        assertEquals("shoes", state.category)
        assertEquals(emptyList(), state.subcategories)
    }

    @Test
    fun `a suggested type fills the seasons in, but never over a choice`() {
        val implied = form().withSuggestedType("shoes", "Sandals", ::seasonsForSubcategories)
        assertEquals(listOf(Season.SUMMER), implied.seasons)

        val chosen = form().copy(seasons = listOf(Season.WINTER))
            .withSuggestedType("shoes", "Sandals", ::seasonsForSubcategories)
        assertEquals(listOf(Season.WINTER), chosen.seasons, "an explicit choice was overwritten")
    }

    @Test
    fun `an import fills an empty brand and leaves a typed one alone`() {
        val empty = form().withImportedPreview(listOf("i1.jpg", "i2.jpg"), "Imported")
        assertEquals(listOf("i1.jpg", "i2.jpg"), empty.imageUris)
        assertEquals(listOf("", ""), empty.bgRemovedUris)
        assertEquals(0, empty.selectedImageIndex)
        assertEquals("Imported", empty.brand)

        val typed = form().copy(brand = "Mine").withImportedPreview(listOf("i1.jpg"), "Imported")
        assertEquals("Mine", typed.brand, "an import overwrote work in progress")

        // An import with no brand does not blank a typed one either.
        assertEquals("Mine", form().copy(brand = "Mine").withImportedPreview(listOf("i1.jpg"), null).brand)
    }

    @Test
    fun `the gallery shows the cut-out and remembers the original`() {
        val items = form(listOf("a.jpg", "b.jpg"), listOf("a-cut.png", "")).galleryItems()

        assertEquals(listOf("a-cut.png", "b.jpg"), items.map { it.uri })
        assertEquals(listOf("a.jpg", "b.jpg"), items.map { it.original })
    }

    @Test
    fun `the preview shows whatever the selected slot has`() {
        assertEquals("a-cut.png", form(listOf("a.jpg"), listOf("a-cut.png")).displayedPreviewUri())
        assertEquals("a.jpg", form(listOf("a.jpg")).displayedPreviewUri())
        assertEquals(null, form().displayedPreviewUri())
    }

    @Test
    fun `a garment imported as a cut-out has nothing to undo to`() {
        // Both slots hold the same path, so there is no with-background original
        // to restore and nothing to re-run removal against.
        val imported = form(listOf("cut.png"), listOf("cut.png"))
        assertTrue(!imported.selectedHasOriginal())

        assertTrue(form(listOf("a.jpg"), listOf("a-cut.png")).selectedHasOriginal())
        assertTrue(!form().selectedHasOriginal())
    }

    @Test
    fun `brand suggestions are what the list is for`() {
        val known = listOf("Adidas", "Arket", "Nike", "Uniqlo")

        // An empty field offers everything: the list doubles as a picker.
        assertEquals(known, brandSuggestions(known, ""))
        // Matching is case-insensitive and anywhere in the name.
        assertEquals(listOf("Adidas", "Arket"), brandSuggestions(known, "a"))
        assertEquals(listOf("Uniqlo"), brandSuggestions(known, "QLO"))
        // An exact match is dropped: it has already been typed.
        assertEquals(emptyList(), brandSuggestions(listOf("Nike"), "nike"))
        // And the list is capped, so it cannot cover the form.
        assertEquals(
            GarmentFormState.BRAND_SUGGESTION_LIMIT,
            brandSuggestions((1..20).map { "Brand $it" }, "").size,
        )
    }
}
